account-kit/infra/src/middleware/gasManager.ts

```diff
@@ -188,6 +212,133 @@ export function alchemyGasAndPaymasterAndDataMiddleware(
           : {}),
       });
 
+      let erc20Context:
+        | {
+            tokenAddress: Address;
+            maxTokenAmount?: bigint;
+            permit?: Hex;
+          }
+        | undefined = undefined;
+      if (policyToken !== undefined) {
+        let maxAmountToken = policyToken.maxTokenAmount
+          ? policyToken.maxTokenAmount
+          : maxUint256;
+
+        erc20Context = {
+          tokenAddress: policyToken.address,
+          ...(policyToken.maxTokenAmount
+            ? { maxTokenAmount: maxAmountToken }
+            : {}),
+        };
+        if (policyToken.approvalMode === "PERMIT") {
+          // get a paymaster address
+          let paymasterAddress: Address | undefined = undefined;
+          const paymasterData = await (
+            client as AlchemySmartAccountClient
+          ).request({
+            method: "pm_getPaymasterStubData",
+            params: [
+              userOp,
+              account.getEntryPoint().address,
+              toHex(client.chain.id),
+              {
+                policyId: Array.isArray(policyId) ? policyId[0] : policyId,
+              },
+            ],
+          });
+
+          paymasterAddress = paymasterData.paymaster
+            ? paymasterData.paymaster
+            : paymasterData.paymasterAndData
+            ? (paymasterData.paymasterAndData.slice(0, 42) as Address)
+            : undefined;
+
+          if (paymasterAddress === undefined) {
+            throw new Error("no paymaster contract address available");
+          }
+          const deadline = maxUint256;
+          const eip712Abi = [
+            "function nonces(address owner) external view returns (uint)",
+          ];
+          const { data } = await client.call({
+            to: policyToken.address,
+            data: encodeFunctionData({
+              abi: parseAbi(eip712Abi),
+              functionName: "nonces",
+              args: [client.account?.address],
+            }),
+          });
+          if (!data) {
+            throw new Error("No nonces returned from erc20 contract call");
+          }
+          console.log(data);
+          const typedPermitData = {
```

Comment:
curious if you have considered extracting these types (& maybe the abi?) outside of this function to make it a little more succinct?

Reply:
yes extracted to other files